lib: os: reboot: a few cleanups

gmarull opened the pull request:
After https://github.com/zephyrproject-rtos/zephyr/commit/62dbe72cb7ac55d1ae3ce62e5fb1fae47b9b293a and discussions in https://github.com/zephyrproject-rtos/zephyr/pull/60626 I realized the reboot functionality required a bit of love. So I tried to clean up things a little bit. I may have made a few mistakes, so please, review. I tried to make changes in small steps to ease the review process.

Fixes #60915

gmarull committed:
lib: os: reboot: remove redundant cache ifdeffery

sys_cache* calls are no-ops when cache options are not enabled.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lib: os: reboot: use CODE_UNREACHABLE

Instead of printking and looping.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lin: os: reboot: fix kernel/irq includes

Module is using irq.h API, not kernel.h APIs.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lib: os: reboot: guard all reboot hooks with CONFIG_REBOOT

If CONFIG_REBOOT is not enabled, there's no need to compile reboot
hooks.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: arm: renesas_smartbond: remove redundant reboot hook

NVIC_SystemReset() is what ARM default hook does.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lib: os: reboot: remove empty reboot hooks

An empty reboot hook violates the implementation requirements, as the
function will return. CONFIG_REBOOT can only be enabled for platforms
supporting it.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: watchdog: counter: do not use reboot hooks

Reboot hooks are reserved for sys_reboot(). Also replace `0` with its
corresponding definintion: SYS_REBOOT_WARM.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lib: os: reboot: improve Doxygen documentation

Reboot module had missing Doxygen docs (e.g. for reboot types) or
function parameters.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lib: os: reboot: enumerate reboot modes

Instead of using defines.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lib: os: reboot: improve reboot mode documentation

Improve the description of reboot modes, to make things a bit more clear.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lib: os: reboot: remove redundant FUNC_NORETURN

It's already provided in the declaration.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lib: os: reboot: move sys_arch_reboot declaration to sys/reboot.h

Also add documentation for it, making it clear what are the API
requirements.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lib: os: reboot: s/sys_arch_reboot/z_sys_reboot

Rename the reboot hook to z_sys_reboot. `z` prefix makes it clear that
function is internal (not meant to be called by application or drivers
code). Removal of `arch` makes sense because it looks like reboot
operation is not always arch specific.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lib: os: reboot: remove redundant reboot hook docstrings

Interface is documented in sys/reboot.h

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
arch: x86: core: reboot: simplify implementation

Simplify the implementation by always performing a reboot (otherwise
function is a no-op, so violates sys_reboot() requirements).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
modules: tfm: remove redundant ifdeffery

CONFIG_TFM_PARTITION_PLATFORM is already used to decide wether to
compile the file or not.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
modules: tfm: reboot: fix includes

kernel.h API was not used, but reboot/toolchain instead.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lib: os: reboot: extend usage of SYS_REBOOT_DEFAULT

There was a wide variety of reboot types in-tree, likely because the
interface was not documented properly nor specified any details. All
code without enough platform-specific knowledge should use
SYS_REBOOT_DEFAULT.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lib: os: reboot: introduce HAS_REBOOT

Add a new hidden option to be selected when an actual reboot
implementation is present. This prevents selection of CONFIG_REBOOT if
not supported.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
lib: os: reboot: guard reboot call and fix Kconfig usage

Modules cannot select REBOOT, because the functionality may not always
be available and so selectable. If it is a strict requirement, depends
on can be used, or imply in some others. This means that generic code
can only attempt reboot if enabled. This is tipically needed in generic
code that can run on platforms without reboot.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: arm: rt6xx: implement custom reboot hook

It looks like extra magic is required to reboot on rt6xx SoCs. Right
now, this was part of board code, let's move it to a custom hook in SoC
code.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: pm_cpu_ops: implement PSCI reboot hook

62dbe72 introduced yet another reboot API for no reason. This patch
converts the implementation into a sys_reboot() hook. Note that shell
commands have been deleted, as there's already a shell command that uses
the standard reboot.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: arm: ti_simplelink: cc32xx: do not rely on enum value

It is unsafe to rely on enumeration values, perform equality check
instead. Also add a loop to wait for reset.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

nashif commented:
The cleanup lgtm, however, I would put the fix for https://github.com/zephyrproject-rtos/zephyr/issues/60915 into its own PR

gmarull committed:
arch: arm: aarch32: nmi: remove reboot dependency

Runtime NMI does not use the reboot API.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: riscv-privileged: virt: cleanup implementation

- Include what is needed
- Remove dead definitions
- Use sys_write32

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: riscv|xtensa: espressif_esp32: esp_restart_noos does not return

Add FUNC_NORETURN attribute.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
arch: arm: cortex_m: add CODE_UNREACHABLE after NVIC_SystemReset()

NVIC_SystemReset() spins forever, so it does not return. However, it is
not flagged with a noreturn attribute.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: riscv-privileged: telink_b91: wait for reboot

Make sure function never returns.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: arm: ti_lm3s6965: wait for reboot

Make sure function never returns.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
modules: tfm: reboot: add CODE_UNREACHABLE

tfm_platform_system_reset() should never return, so add
CODE_UNREACHABLE.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

de-nordic commented on subsys/mgmt/mcumgr/grp/os_mgmt/src/os_mgmt.c:
```diff
@@ -310,7 +310,7 @@ static int os_mgmt_taskstat_read(struct smp_streamer *ctxt)
 static void os_mgmt_reset_work_handler(struct k_work *work)
 {
-	sys_reboot(SYS_REBOOT_WARM);
+	sys_reboot(SYS_REBOOT_COLD);
```
@nordicjm Do we specify type for reboot?

gmarull replied:
Note: the meaning of cold/warm reboot was not clear (there were no docs at all), so it was used randomly. The meaning of "warm" reboot is still not much clearer (it's probably driven by PSCI warm reboot).

gmarull replied:
> This function is/was a bit of a lie, WARM/COLD is not what this function has done, warm reboot _should_ be jumping to the restart vector of the application, cold reboot _should_ be using the CMSIS restart function, but actually what this does is CMSIS restart in both cases but (depending on a Kconfig) sets one of the GPREGRET registers to a value. Nothing in-tree uses it and neither does nor has MCUboot ever used it, it should never have been added and was not fit for purpose.

Note that this is your interpretation of warm/cold because it was not documented at all.

nordicjm replied:
warm/cold reboot are standard terms, you would expect a cold reboot to reset the hardware e.g. watchdog timer, you would not expect a warm reboot to do that

de-nordic replied:
cold reboot is basically what you expect to happen on powering up a device; warm reset may be just jumping at specified vector that will reinitialize the software you are warm rebooting, which means that all set-once-per-runtime things, like some hardware assisted trusted enclaves and so on, may persist in state they have been set by powerup.

nashif replied:
one of those things defined in simple terms on wikipedia: https://en.wikipedia.org/wiki/Reboot. We end up having this discussion everytime the API is touched. This API has been with us since Zephyr was launched and if memory serves right, it was mostly used on x86 systems, but then got adopted by various platforms. Warm and Cold are probably very limited in covering the various use-cases we have these days.

Maybe it time to deprecate this and move to something a bit more suitable to all platforms and usecases we have now.

gmarull replied:
According to wikipedia description, we’ve never had cold reboot then, because it implies “power off” as well. The main problem is that Zephyr never defined what its reboot types mean, and as a result, implementations differ and there’s a mix of warm/cold usage because nobody knows what to choose.

nashif replied:
hmm, this what we had in Zephyr v1.0.0:

```
static inline void cold_reboot(void)
{
        uint8_t reset_value = SYS_X86_RST_CNT_CPU_RST | SYS_X86_RST_CNT_SYS_RST |
                                                        SYS_X86_RST_CNT_FULL_RST;
        sys_out8(reset_value, SYS_X86_RST_CNT_REG);
}

void sys_arch_reboot(int type)
{
        switch (type) {
        case SYS_REBOOT_COLD:
                cold_reboot();
                break;
        default:
                /* do nothing */
                break;
        }
}
```

So, never is not accurate, unless this code does not do what was intended.

de-nordic replied:
> According to wikipedia description, we’ve never had cold reboot then, because it implies “power off” as well. The main problem is that Zephyr never defined what its reboot types mean, and as a result, implementations differ and there’s a mix of warm/cold usage because nobody knows what to choose.

It does not have to be power off. Cold reset on old AT and PS/2 platforms was controlled by external 8042 chip that could be either triggered by software or short-circuited by reset pin on the device enclosure.

Many MCUs have external MCLR/RESET pin that can be controlled by external device, and works similarly to power cycle, as it is often intended to be controlled by specialized Reset IC. I have used a reset circuit where MCU, not running Zephyr (sorry),  would trigger external reset circuit doing t1-wait->t2-hold->release for itself and other MCUs with all MCLR connected to the same line, basically power cycling itself and others.

The MCUs get more and more complicated and have internal DC/DC converters and power rails that can be disabled and enabled by PM controller for various reasons; if there is an MCU with PM controller where you can trigger it to power down all rails for a moment that would be a cold reset, from the software perspective.

Shouldn't we have PM  experts from IC vendors, contributing to Zephyr, gathered in single room and have them discuss what they have and how can we address it?

gmarull replied:
> Shouldn't we have PM experts from IC vendors, contributing to Zephyr, gathered in single room and have them discuss what they have and how can we address it?

That would be ideal, but... do we have access to such expertise?

aescolar replied:
The thing is, with the kind of µCs Zephyr supports in general, there is a lot of possible "reset" options.
Starting with: Are you resetting just the MCU which is executing that code? (maybe with its closest linked peripherals, maybe its whole "domain"?) the whole SOC? more ICs than just the SOC?
Then even if you may be resetting "the whole SOC", that SOC may easily support different reset modes (will it reset GPIOs configuration, security enclaves,..), and the SOC may not allow all resets to be triggered from any states or from any of its µCs.
So I would lean to leave this quite open and overridable by the SOC/board. Even if at some point some kind of resets being so common as to be explicitly listed at the kernel level, I'd still let SOCs and boards expand that list.

gmarull replied:
The problem is that if we open `sys_reboot()` to vendor-specific reset options, the API becomes even less portable. Refer to https://github.com/zephyrproject-rtos/zephyr/pull/60920/commits/a7e841ba31842295dfcfcb76503f0b02557f13d1. Remember this API is called from generic code that has no specific system knowledge.

aescolar replied:
> The problem is that if we open sys_reboot() to vendor-specific reset options, the API becomes even less portable.

Well.. not necessarily. At Zephyr's level we'd define the basic reset modes any code can call.
If a ARCH or SOC family supports some extra reset modes, it would only be then code from that ARCH/SOC family (or its drivers) or apps which specifically target those who would call with that reset type.
We also specify that a not supported reset mode is received, we fall back to another type of reset.

aescolar commented on arch/posix/core/cpuhalt.c:
```diff
@@ -45,17 +43,3 @@ void arch_cpu_atomic_idle(unsigned int key)
- *
- * Does nothing
- */
-void __weak sys_arch_reboot(int type)
```
I may be missing something, but instead of removing this, shouldnt be select HAS_REBOOT from ARCH_POSIX, and keep this (now named `z_sys_reboot()`)

gmarull replied:
I guess it's ok to allow posix implementation to shut the application down because it cannot reboot, then? A reboot in posix would be like `sys_reboot()` + relaunch zephyr manually?

aescolar replied:
Yep, this default reboot hook is just terminating returning 1 to the calling program after printing a descriptive message. A board, or app developer can override that with something else.
Normally a reboot is the result of a failure. And in an environment like this, terminating by default is reasonable.

gmarull replied:
please check latest revision. I've removed __weak (non-standard feature_ in favor of a Kconfig option (like ARM)

aescolar reviewed (dismissed):
see previous comment

gmarull committed:
arch: posix: fix documentation for reboot implementation

Implementation exits the application, so "does nothing" was incorrect.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
arch: posix: rephrase reboot message

Usage of __func__ exposes internal details (hook name), this is not
needed.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

aescolar commented on include/zephyr/sys/reboot.h:
```diff
@@ -20,19 +13,48 @@
+	SYS_REBOOT_COLD,
+	/** Warm reboot. */
+	SYS_REBOOT_WARM,
+};
```
Not trying to open an endless discussion, but: 
> Cold reboot is basically what you expect to happen on powering up a device;

Agreed.

> warm reset may be..

..a lot of different things, and for HW assisted reboots triggered by SW, the HW may provide several options with significantly different outcomes.
Which makes me wonder if Zephyr should not make it easier for platforms to define their own different reboot types. This is probably out of scope for this PR, though this enum seems to be tying things up?

gmarull replied:
I think it's easy to agree on what cold boot means, but not so much on "warm". I'm not even sure we're now able to support something like warm reset properly, see, for example, https://trustedfirmware-a.readthedocs.io/en/latest/design/reset-design.html

reboot types have been undocumented from day 0 https://github.com/zephyrproject-rtos/zephyr/commit/826edfc16b2615e8ee5f8fd0b06af39ded365a8e, who knows what they meant in 2016...

gmarull committed:
soc: posix: add FUNC_NORETURN to posix_exit

Function does not return.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

ceolin commented on soc/arm/nxp_s32/common/power_soc.c:
```diff
@@ -15,49 +15,23 @@
-		.McRgmConfigPtr = (const Power_Ip_MC_RGM_ConfigType *)&mc_rgm_cfg,
-		.PMCConfigPtr = NULL
+	Power_Ip_MC_RGM_ConfigType mc_rgm_cfg = {
+		.ResetType = MCU_FUNC_RESET,
```
This target had different COLD/WARM implementations, my understanding is that you are dropping WARM usage across the tree, so why here you ignoring the given parameter and arbitrary using the WARM path ? maybe a typo ?

COLD was `mc_rgm_cfg.ResetType = MCU_DEST_RESET;` 
WARM `mc_rgm_cfg.ResetType = MCU_FUNC_RESET;`

gmarull replied:
this is done on purpose. `sys_reboot()` is not expected to perform destructive actions, maybe some new APIs should be introduced https://github.com/zephyrproject-rtos/zephyr/pull/60920/commits/d4b0d71107f72effbcce02cd66ada2fa01a0b6b8

If we allow any kind of behavior for sys_reboot() there's no way to write portable code.

ceolin replied:
Looking the rest of the changes it seems it should be:
```
if (typ == SYS_REBOOT_WARM) {
		mc_rgm_cfg.ResetType = MCU_FUNC_RESET;
} else {
		mc_rgm_cfg.ResetType = MCU_DEST_RESET;
}

Power_Ip_PerformReset(&power_cfg);

for (;;) {
		/* wait for reset */
}
```

ceolin replied:
> this is done on purpose. `sys_reboot()` is not expected to perform destructive actions, maybe some new APIs should be introduced [d4b0d71](https://github.com/zephyrproject-rtos/zephyr/commit/d4b0d71107f72effbcce02cd66ada2fa01a0b6b8)
> 
> If we allow any kind of behavior for sys_reboot() there's no way to write portable code.

I am pretty confuse now, isn't cold reboot something that can wipe out your ram, cut the power, ... basically any context is lost ? This target was "supposedly" implementing it. If that is not the case, what it is cold and warm boot for us ? I honestly don't want an infinite discussion, these changes are good, what I don't understand is why this target that implements both WARM / COLD reboots  have the COLD path dropped ... Is that because of other changes in the cold replacing the WARM with COLD ?

If that is the case, we may re-evaluate why having two types of reboot

manuargue replied:
@gmarull I'm okay with the change if `sys_reboot` it's not supposed to perform destructive actions. At the time I did this, it was not clear to me what is the expectation of cold/warm reset, hence the description I've included in this function.

gmarull replied:
Thanks for the review @manuargue. Do you know under which circumstances the destructive reset is used/preferred?

manuargue replied:
> Thanks for the review @manuargue. Do you know under which circumstances the destructive reset is used/preferred?

@gmarull In general for the S32 devices, the request sequence goes like `POR -> destructive reset sequence -> functional reset sequence -> chip out of reset -> perform boot flow`.
Destructive reset (either coming from POR or from an event that triggered it) will reset everything except the POR and debug logic, so this is the closest to POR.
Functional reset, will not reset all modules (e.g. RTC, PMC, fault detections) and SRAM content is preserved.

nordicjm replied:
With the destructive reset, that would cause some issues with some zephyr systems e.g. retained RAM. I agree this should be standardised as to what a reset does and preserves.

gmarull replied:
@manuargue so, which reset should be the default in this case? is destructive always preferred or not?

manuargue replied:
+1 for standardization. In general, reset is platform dependent, but we could define what is expected to be reset/preserved for each of the types of reset defined by Zephyr. Otherwise we leave the door open to interpretations and inconsistent behaviors between SoCs.

gmarull replied:
I think it'll be difficult to standardize the reboot modes because there seems to be a wide number of variants in the field. I'd tend to leave warm/cold options, but let implementations decide what do they mean exactly. We can ask for certain guarantees, e.g. warm resets less than cold, it's faster, etc. But I don't think we can expect to write portable code that uses `SYS_REBOOT_COLD/WARM`.

This is the reason `SYS_REBOOT_DEFAULT` https://github.com/zephyrproject-rtos/zephyr/pull/60920/commits/8ed86746da92684b9723068a21ff98fa14962486 has been introduced. If some portable code has no clue on which reset mode to choose, it should provide `SYS_REBOOT_DEFAULT` so that the system decides the _most suitable_ one if not specified.

Also, for reference, it looks like Linux does allow setting distinct reset modes only via Kernel boot parameters https://elixir.bootlin.com/linux/latest/source/kernel/reboot.c#L1068. So it seems like it's something that one consciously decides when booting the Kernel, and so likely knowing the implications of reset modes. `SYS_REBOOT_DEFAULT` goes in line with this approach.

Btw, @manuargue can you review your -1?

ceolin commented on arch/x86/core/reboot_rst_cnt.c:
```diff
@@ -20,21 +21,14 @@
-				X86_RST_CNT_FULL_RST;
-	sys_out8(reset_value, X86_RST_CNT_REG);
-}
+	ARG_UNUSED(type);
```
It is probably better to add an assert here instead of just change the implementation. It may help catch possible wrong usage.
`__ASSERT(type == SYS_REBOOT_COLD)`

gmarull replied:
the API contract has been refined, please check https://github.com/zephyrproject-rtos/zephyr/pull/60920/commits/dfb5e2508664dcd7d7ae9c9ff9bae751d7768971

ceolin replied:
Looked it, doesn't it make sense to at least log on targets that were doing distinct things before ?

ceolin replied:
anyway, that is minor.

carlocaione commented on drivers/pm_cpu_ops/pm_cpu_ops_psci.c:
```diff
@@ -98,27 +104,22 @@ static int psci_features_check(unsigned long function_id)
 
-	return psci_to_dev_err(ret);
-
+	for (;;) {
+		/* wait for reboot */
+	}
```
I understand we cannot deal with the return value from the PSCI call anymore, but can we at least get that anyway and maybe print an error if the returned value is not the one expected?

gmarull replied:
`sys_reboot()` has never returned a value, so we could assert or log. We could also consider allowing `sys_reboot()` to return errors, and only guarantee a no return if successful. Let's discuss later today in arch wg.

carlocaione replied:
yes, I was referring to assert or log. If then we can make `sys_reboot()` to return even better to me.

gmarull replied:
added a printk

manuargue commented on soc/arm/nxp_s32/s32ze/Kconfig.series:
```diff
@@ -14,5 +14,6 @@ config SOC_SERIES_S32ZE_R52
 	select VFP_DP_D16
 	select PLATFORM_SPECIFIC_INIT
 	select SOC_FAMILY_S32
+	select ARM_CUSTOM_REBOOT
```
pls remove this line, custom reboot not yet supported on this soc, building `nxp_s32/common/power_soc.c` is guarded for now to build for s32k only.

gmarull replied:
thanks, fixed

gmarull committed:
soc: arm: nxp_s32: always perform a reboot

Reboot hook must not return, default to functional (non-destructive)
reset mode. Also add a wait loop at the end.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: riscv: openisa_rv32m1: wait for reset

Spin to wait for reset to happen.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
Major changes since last revision: https://github.com/zephyrproject-rtos/zephyr/pull/60920/commits/b9ecbfcab85b4c16a6275f3c1cae3da38cb2150d (to be discussed in arch wg today)

carlescufi commented:
Arch WG:

- @gmarull showed the cleanup in general
- @nashif looks at the changes in `sys/reboot.h` and states that we must be careful with changes like those. Also, looking at Linux is not always the right thing since it doesn't translate directly due to the complexity
- @carlescufi proposes keeping the `sys_reboot(enum )` API but adding a new `enum` option that is `DEFAULT`.

gmarull committed:
lib: os: reboot: introduce SYS_REBOOT_DEFAULT

When using this new mode, caller of sys_reboot() lets the system to
device the most appropriate reboot mode (platform dependent). This
option can be used to write portable code, where there is not enough
context information to device the most suitable reboot mode.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
shell: modules: kernel: improve reboot command

Convert reboot into a single command that optionally accepts a reboot
mode. If not provided, SYS_REBOOT_DEFAULT is used.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
last push: introduced `SYS_REBOOT_DEFAULT` as discussed in the call

martinjaeger commented on subsys/task_wdt/Kconfig:
```diff
@@ -5,7 +5,6 @@
 
 menuconfig TASK_WDT
 	bool "Task-level software watchdog"
-	select REBOOT
```
This should be converted to `depends on`. Otherwise a user could enable the task watchdog without enabling `REBOOT`, leading to a no-op when no custom callback is set and the watchdog triggers. This would break previous default behavior.

aescolar commented on include/zephyr/sys/reboot.h:
```diff
@@ -20,19 +13,66 @@
+	/**
+	 * Cold reboot.
+	 *
+	 * This mode is the one preserving the least state possible in the
```
Would it be better to say for the cold reboot something like:
"The system will be rebooted in a way comparable to a power on reset.
What this means depends on the platform, but you can normally expect the device will boot from ROM, and  its peripherals will be re-initialiazed"

And for the warm reboot:
"Typically a faster and lighter reboot than COLD. What this means and what is reset is greatly platform dependent."

gmarull replied:
good point, added these descriptions. thanks!

manuargue reviewed (dismissed):
My comments addressed

gmarull commented:
last push: added a missing CONFIG_REBOOT=y in task_wdt sample

nashif commented:
we still have 2 changes that will possibly break out of tree implementations:
- Moving from weak symbol to requiring HAS_REBOOT
- Change of the name implementing the reboot code, i.e. `sys_arch_reboot()` to `z_sys_reboot()`

gmarull commented:
> we still have 2 changes that will possibly break out of tree implementations:
> 
> * Moving from weak symbol to requiring HAS_REBOOT

weak change is not related to HAS_REBOOT. This PR removes weaks usage in favor of Kconfig options so that non-standard C features are not used. HAS_REBOOT prevents selecting REBOOT if not implemented.

> * Change of the name implementing the reboot code, i.e. `sys_arch_reboot()` to `z_sys_reboot()`

carlescufi commented:
> IMHO stability should be guaranteed for the public interface. If we want to keep stability even for internals, we’ll never evolve.

I agree, and especially for architectures. Out-of-tree architectures are not supported by Zephyr.

nashif commented:
> weak change is not related to HAS_REBOOT. This PR removes weaks usage in favor of Kconfig options so that non-standard C features are not used. HAS_REBOOT prevents selecting REBOOT if not implemented.

that is what I mean, you are removing weak usage in favor of Kconfig (`Moving from weak symbol to requiring HAS_REBOOT`), so an SoC implementation using the weak function and without HAS_REBOOT will break.

> this is done because reboot doesn’t seem to be arch specific only, also to clarify this is an internal API.

It is a hook someone can implement in their platform, This is not a classic internal API. This is more like `arch_` and `soc_` calls we have in various places.

> I agree, and especially for architectures. Out-of-tree architectures are not supported by Zephyr.

Who said anything about out of tree architectures? This functional is usually implemented by platforms or SoCs. Look at this PR, most of the changes are in SOC code implementing this feature.

Anyways, I know most of this is special in many ways, but I do not see anything in this PR documenting or at least trying to help or mitigate breakage. This is one more thing that will break and make migration to a new release more painful for out of tree users.

carlescufi commented:
> > I agree, and especially for architectures. Out-of-tree architectures are not supported by Zephyr.
> 
> Who said anything about out of tree architectures? This functional is usually implemented by platforms or SoCs. Look at this PR, most of the changes are in SOC code implementing this feature.

Fair enough, but the rename makes sense, since this is not arch-specific as you point out.

nashif commented:
> Not really. It is true that now implementers of the hook need to select HAS_REBOOT, if not, CONFIG_REBOOT is not selectable.

which is it? if architecture does not select HAS_REBOOT and I have implemented a custom reboot using the weak function, I end up with my application building with reboot disabled, so this is bad.

if HAS_REBOOT is selected by someone, i.e. architecture, I will get reboot functionality but my application and platform will be using different code than the one I have implemented out of tree, i.e. no warning or complaints and users wondering what went wrong....

> `sys_arch_reboot` was an internal API, never meant to be used by application code. Note that it was not documented, and no declaration existed in any header (reboot.c just declared it internally).

Not application code, correct. but it is a hook a platform can implement. I think this is a class of interfaces we need to differentate from classic internal functions.  Not some I would compare to public APIs and the policy around them, but it should be dealt with differently. Something to discuss somewhere else I guess.

gmarull commented:
> > Not really. It is true that now implementers of the hook need to select HAS_REBOOT, if not, CONFIG_REBOOT is not selectable.
> 
> which is it? if architecture does not select HAS_REBOOT and I have implemented a custom reboot using the weak function, I end up with my application building with reboot disabled, so this is bad.

Architecture will select HAS_REBOOT only if it provides a valid hook. So if it didn't, your code will still work provided you now select HAS_REBOOT in your own hook, regardless of weak usage. HAS_REBOOT is required, yes, but I think it is being added to solve another problem (linking errors for those not implementing reboot). The only problem may come if your arch supported weaks, but now relies on a Kconfig to explicitly choose a custom hook (ARM/POSIX). This is again to avoid usage of non-standard C features. It is a matter of documenting the changes in the release notes, and offering some quick migration tips.
 
> 
> if HAS_REBOOT is selected by someone, i.e. architecture, I will get reboot functionality but my application and platform will be using different code than the one I have implemented out of tree, i.e. no warning or complaints and users wondering what went wrong....
> 
> > `sys_arch_reboot` was an internal API, never meant to be used by application code. Note that it was not documented, and no declaration existed in any header (reboot.c just declared it internally).
> 
> Not application code, correct. but it is a hook a platform can implement. I think this is a class of interfaces we need to differentate from classic internal functions. Not some I would compare to public APIs and the policy around them, but it should be dealt with differently. Something to discuss somewhere else I guess.

As I said, I think internals should not be part of stability guarantees. I'd apply the same for drivers, actually. It is a very different scenario to be a driver interface user than a driver implementer. For example, a driver class may decide to gather 2 driver ops into a single one, but keep the same public interface. If doing oot drivers, you should be prepared to deal with such internal changes, otherwise take some time to upstream your code. In this particular case, the interface wasn't even exposed in any header, nor documented. I think stability concerns are sometimes going too far, making simple changes too difficult.

nashif commented on arch/arm/core/aarch32/Kconfig:
```diff
@@ -257,6 +257,12 @@ config ARM_STORE_EXC_RETURN
 	  This is needed when switching between threads that differ in either
 	  FPU usage or security domain.
 
+config ARM_CUSTOM_REBOOT
```
why  is custom reboot an ARM only config? AFAIK Previously with weak symbols this would have worked for all architectures. I would change this to become a gloabl config that can be used by all platforms, not only ARM.

gmarull replied:
Because not all architectures may need custom hooks. Some architectures will not even have a default hook. If preferred, I can create a global `CONFIG_REBOOT_NO_ARCH_DEFAULT` that can be shared, though

gmarull replied:
note: the goal here is to remove usage of non-standard C features

nashif replied:
> Because not all architectures may need custom hooks. Some architectures will not even have a default hook. If preferred, I can create a global `CONFIG_REBOOT_NO_ARCH_DEFAULT` that can be shared, though

hmm, not sure what `CONFIG_REBOOT_NO_ARCH_DEFAULT`  is about, just move this config outside ARM and make it generic, this way all platforms can override architecture implementation and be happy, I do not see why this needs to be ARM only.

> note: the goal here is to remove usage of non-standard C features

you are referring to _weak symbols? I am fine with that, but we need to keep the implementation compatible here.

gmarull replied:
> > Because not all architectures may need custom hooks. Some architectures will not even have a default hook. If preferred, I can create a global `CONFIG_REBOOT_NO_ARCH_DEFAULT` that can be shared, though
> 
> hmm, not sure what `CONFIG_REBOOT_NO_ARCH_DEFAULT` is about, just move this config outside ARM and make it generic, this way all platforms can override architecture implementation and be happy, I do not see why this needs to be ARM only.

will add a generic symbol

> 
> > note: the goal here is to remove usage of non-standard C features
> 
> you are referring to _weak symbols? I am fine with that, but we need to keep the implementation compatible here.

keeping things compatible and using non-standard C features is incompatible :-)

gmarull committed:
lib: os: reboot: introduce REBOOT_NO_ARCH

Option to signal that reboot functionality is implemented as an
alternative to architecture default code. This option only needs to be
selected if architecture code provides a default reboot implementation
but an alternative implementation is required in platform code.
Architecture code providing a default reboot mechanism must avoid
compiling in the default if this option is selected.

This option allows avoiding usage of non-standard weaks in C.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
doc: release-notes: detail reboot changes

Add details on changes made in the reboot.h API.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
last push: added CONFIG_REBOOT_NO_ARCH (to replace arch weaks in a generic way) and added release notes detailing all changes

manuargue commented on doc/services/retention/index.rst:
```diff
@@ -182,7 +182,7 @@ which will allow rebooting directly into the serial recovery mode by using:
 
 	bootmode_set(BOOT_MODE_TYPE_BOOTLOADER);
-	sys_reboot(0);
+	sys_reboot(SYS_REBOOT_DEFAULT);
```
It will preserve RAM contents since the default is the functional reset, described in https://github.com/zephyrproject-rtos/zephyr/pull/60920#discussion_r1281476723

nordicjm reviewed (dismissed):
Retention/retained memory OK

nordicjm reviewed (approved):
OK for retention/mcuboot

de-nordic reviewed (approved):
MCUmgr/MCUboot look ok.

gmarull commented:
so what is exactly blocking this PR now?

de-nordic commented:
> so what is exactly blocking this PR now?

Four days waiting and already inpatient.

gmarull commented:
> > so what is exactly blocking this PR now?
> 
> Four days waiting and already inpatient.

not impatient, just unsure on why it's being blocked.

fabiobaltieri commented:
@gmarull I think I recall @nashif mentioning that this should go through the treewide process https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes

gmarull commented:
Closing this one, if someone else has an interest and patience in this, feel free to take it.

martinjaeger commented:
It's a shame this very useful cleanup didn't make it in.

@nashif It looks like your previous concern has been addressed. Is there still something else blocking?

nashif commented:
> @nashif It looks like your previous concern has been addressed. Is there still something else blocking?

this is a [treewide](https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes) change, which changes behaviour and touches many platforms and other areas in the tree that needs to go the through the process defined [here](https://docs.zephyrproject.org/latest/contribute/guidelines.html#requirements-for-treewide-changes).
